File: Leetcode-Medium/15-3Sum/3Sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        ans = []
        n = len(nums)
        nums.sort()
        for i in range(n):
            front = i+1
            rear = n-1
            if rear-front==0:
                break
            sum = 0-nums[i]
            while(front<rear):
                if nums[front]+nums[rear]==sum:
                    ans.append([nums[i],nums[front],nums[rear]])
                    temp1 = nums[front]
                    temp2 = nums[rear]
                    while(temp1==nums[front] and front<rear):
                        front+=1
                    while(temp2==nums[rear] and front<rear):
                        rear-=1
                elif nums[front]+nums[rear]<sum:
                    front+=1
                else:
                    rear-=1
        new_ans = []
        for elem in ans:
            if elem not in new_ans:
                new_ans.append(elem)
        ans = new_ans
        return ans

File: Leetcode-Medium/15-3Sum/3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        if(nums.size()<3)return {};
        sort(nums.begin(),nums.end());
        int n =nums.size();
        if(nums[0]>0)return {};
        map<int,int> m ;
        for(int i=0;i<nums.size();i++)
            m[nums[i]]=i;
        vector<vector<int>>ans;
        for(int i=0;i<n-2;i++){
        if(nums[i]>0)break;
            for(int j=i+1;j<n-1;j++){
                int reqd=-1*(nums[i]+nums[j]);
                if(m.count(reqd)&&m.find(reqd)->second>j)
                    ans.push_back({nums[i],nums[j],reqd});
                    j=m.find(nums[j])->second;
            }
            i=m.find(nums[i])->second;
        }
        return ans;
    }
};
